pkg/translate/translateValue.go

```diff
@@ -0,0 +1,554 @@
+	outputMapping := make(map[string]*Translation)
+	for valKey, componentName := range t.ValuesToComponentName {
+		featureName := name.ComponentNameToFeatureName[componentName]
+		if featureName == "" {
+			continue
+		}
+		cnEnabled, err := name.IsComponentEnabledFromValue(valKey, valueTree)
+		if err != nil {
+			log.Errorf("Error checking component:%s enablement, skip k8s mapping", componentName)
+			continue
+		}
+		if !cnEnabled {
+			log.Infof("Component:%s disabled, skip k8s mapping", componentName)
+			continue
+		}
+		for K8SValKey, outPathTmpl := range t.KubernetesMapping {
+			newKey := renderComponentName(K8SValKey, valKey)
+			newVal := renderFeatureComponentPathTemplate(outPathTmpl.outPath, featureName, componentName)
+			outputMapping[newKey] = &Translation{newVal, nil}
+		}
+	}
+
+	t.KubernetesMapping = outputMapping
+}
+
+// NewValueYAMLTranslator creates a new ValueYAMLTranslator for minorVersion and returns a ptr to it.
+func NewValueYAMLTranslator(minorVersion version.MinorVersion) (*ValueYAMLTranslator, error) {
+	t := ValueYAMLTranslators[minorVersion]
+	if t == nil {
+		return nil, fmt.Errorf("no value.yaml translator available for version %s", minorVersion)
+	}
+
+	err := t.initAPIMapping(minorVersion)
+	if err != nil {
+		return nil, fmt.Errorf("error initialize API mapping: %s", err.Error())
+	}
+	return t, nil
+}
+
+// TranslateFromValueToSpec translates from values struct to IstioControlPlaneSpec
+func (t *ValueYAMLTranslator) TranslateFromValueToSpec(values *v1alpha2.Values) (controlPlaneSpec *v1alpha2.IstioControlPlaneSpec, err error) {
+	// marshal value struct to yaml
+	valueYaml, err := yaml.Marshal(values)
+	if err != nil {
+		return nil, fmt.Errorf("error when marshalling value struct %v", err.Error())
+	}
+	// unmarshal yaml to untyped tree
+	var yamlTree = make(map[string]interface{})
+	err = yaml.Unmarshal(valueYaml, &yamlTree)
+	if err != nil {
+		return nil, fmt.Errorf("error when unmarshalling into untype tree %v", err.Error())
+	}
+
+	outputTree := make(map[string]interface{})
+	err = t.TranslateTree(yamlTree, outputTree, nil)
+	if err != nil {
+		return nil, err
+	}
+	outputVal, err := yaml.Marshal(outputTree)
+	if err != nil {
+		return nil, err
+	}
+
+	var cpSpec = &v1alpha2.IstioControlPlaneSpec{}
+	err = util.UnmarshalWithJSONPB(string(outputVal), cpSpec)
+
+	if err != nil {
+		return nil, fmt.Errorf("error when unmarshalling into control plane spec %v", err.Error())
+	}
+
+	return cpSpec, nil
+}
+
+// TranslateTree translates input value.yaml Tree to ControlPlaneSpec Tree
+func (t *ValueYAMLTranslator) TranslateTree(valueTree map[string]interface{}, cpSpecTree map[string]interface{}, path util.Path) error {
+	// translate enablement and namespace
+	err := t.setEnablementAndNamespacesFromValue(valueTree, cpSpecTree)
+	if err != nil {
+		return fmt.Errorf("error when translating enablement and namespace from value.yaml tree: %v", err.Error())
```

Comment:
no need for err.Error(), here and throughout. the only exception is log.Errorf() which expects a string.

Reply:
done